### PR TITLE
Improvements to the strategy implemented by `{fixed,standard,zip317}::SingleOutputChangeStrategy`

### PR DESCRIPTION
**Pull request by daira**
`zcash_client_backend::{fixed,standard,zip317}::SingleOutputChangeStrategy` now implement a different strategy for choosing whether there will be any change, and its value. The aims are:

* Ensure that it is still possible to create fully transparent transactions with no change (this will be needed for ZIP 320). The `ChangeError::InsufficientFunds` error in this case should have a `required` field that reflects the additional amount needed, according to the fee calculated without an extra change output.
* Avoid leaking information about note amounts in some cases: an adversary that knew the number of external recipients and the sum of their outputs was able to learn the sum of the inputs if no change output was present.
* Defend against losing money by using `DustAction::AddDustToFee` with a too-high dust threshold.
* Ensure that if a "change memo" is requested, there will always be a shielded change output in which to put it. Previously, this would not be the case when using `DustAction::AddDustToFee`.

### CODE REVIEW

File: zcash_client_backend/src/fees/zip317.rs [hunk 1]
```diff
@@ -351,13 +350,149 @@ mod tests {
             ),
             #[cfg(feature = "orchard")]
             &orchard_fees::EmptyBundleView,
+            &DustOutputPolicy::new(DustAction::AllowDustChange, Some(NonNegativeAmount::ZERO)),
```
Comment (str4d) [inline zcash_client_backend/src/fees/zip317.rs:353]: Why was the `change_with_transparent_payments` test altered to use this dust output policy (instead of `DustOutputPolicy::default()`)? AFAICT the test is not transparent-only (it has a Sapling input), and therefore the change to the fee code should not have affected it.

If this isn't a necessary change, revert it. If we don't have a test with `DustAction::AllowDustChange`, then we should add that separately.

File: zcash_client_sqlite/src/testing/pool.rs [hunk 2]
```diff
@@ -443,7 +453,7 @@ pub(crate) fn send_multi_step_proposed_transfer<T: ShieldedPoolTester>() {
 
     assert_eq!(
         confirmed_sent.get(0).and_then(|v| v.get(0)),
-        Some(&(&txids[0], 50000))
+        Some(&(&txids[0], 35000))
```
Comment (str4d) [inline zcash_client_sqlite/src/testing/pool.rs:456]: It took me a while to realise that this edit was not due to changes in the existing output amount or how fees were changed, but just checking the value of the new change output instead.

This test now has an undocunmented dependency on implementation details about the order in which the two outputs of the first transaction are added to the database. Change the test to instead sort the outputs by value and check both of them (the existing 50,000 zatoshi non-change output, and the new 35,000 zatoshi change output).

File: zcash_client_sqlite/src/testing/pool.rs [hunk 3]
```diff
@@ -904,7 +914,7 @@ pub(crate) fn spend_succeeds_to_t_addr_zero_change<T: ShieldedPoolTester>() {
     let dfvk = T::test_account_fvk(&st);
 
     // Add funds to the wallet in a single note
-    let value = NonNegativeAmount::const_from_u64(60000);
+    let value = NonNegativeAmount::const_from_u64(70000);
```
Comment (str4d) [inline zcash_client_sqlite/src/testing/pool.rs:917]: Why is this amount increased?

File: zcash_client_sqlite/src/testing/pool.rs [hunk 4]
```diff
@@ -951,7 +961,7 @@ pub(crate) fn change_note_spends_succeed<T: ShieldedPoolTester>() {
     let dfvk = T::test_account_fvk(&st);
 
     // Add funds to the wallet in a single note owned by the internal spending key
-    let value = NonNegativeAmount::const_from_u64(60000);
+    let value = NonNegativeAmount::const_from_u64(70000);
```
Comment (str4d) [inline zcash_client_sqlite/src/testing/pool.rs:964]: Why is this amount increased?

**Review by str4d — CHANGES_REQUESTED** [review]

**Review by str4d — APPROVED** [review]
utACK 977e6c802a0a530c46ad1608cc8a3594c7a53d19 modulo questions about tests. Any that are not addressed here, should have issues opened.

File: zcash_client_backend/src/fees/zip317.rs (hunk 1 above)
Comment (daira) [reply zcash_client_backend/src/fees/zip317.rs:353]: The behaviour intentionally changes in the situation of the original test. Specifically, one of the dummy outputs that was originally present would become a change output instead, because we provided a change memo. (This would be observable to someone with the corresponding internal IVK, although not otherwise.) However, the default dust policy disallows creating a transaction with zero-valued change (the default dust threshold is the ZIP 317 marginal fee of 5000 zat), and so the test would fail without changing the dust policy.

I don't know whether this is the intent of the default policy, but it is strictly speaking what we asked for: we specified a change memo and said to reject if the change is less than the marginal fee. So what I did to make the test pass is explicitly specify a policy that allowed this case (which might not be the correct solution).

In more detail:

The original test has a Sapling input of 55000 zat (so, the flows are not fully transparent), and a transparent output of 40000 zat. The selected change pool is Sapling. There is no explicitly specified Sapling output (and the `bundle_required` setting of the builder defaults to false), but there is at least one Sapling input and so the outputs are padded to 2; [this is the intended behaviour of the Sapling builder](https://github.com/zcash/sapling-crypto/blob/d23547f0d3c753b706c6a4fd64456855e153d6b7/src/builder.rs#L36..L37). There is no Sapling change output. Therefore the ZIP 317 fee is, correctly, 15000 zat because we have 3 logical actions: two due to Sapling (max of one input and two dummy outputs), and one transparent output.

Before the latest changes from our pairing, the PR was incorrect because it was calculating the fee as 20000 zat.

The new behaviour is that we create a zero-valued change output with `change_memo`. Because of the output padding, this makes no difference to the fee of 15000 zat, and so the balance is calculated as 55000 - 40000 - 15000 = 0, i.e. we are in case 2, but the flows are not transparent so this becomes case 3b. However, the default dust policy disallows the change output. Therefore, we reject the transaction as having insufficient funds.

I'm open to either:
* renaming the test;
* having this test explicitly fail and adding another test that allows zero-valued change;
* changing the interpretation of the default dust policy to allow this case. (Should it also allow change less than the marginal fee in general if a change memo is given, or only zero-valued change?)

Comment (daira) [reply zcash_client_backend/src/fees/zip317.rs:353]: An argument for allowing zero-valued change but not allowing other change values less than the threshold, is that we will never need to spend a zero-valued note (and in principle shouldn't be keeping track of its witness, although I don't know whether that is currently the case). And this case may occur in practice for the intended "send max" feature in Zashi.

Comment (str4d) [reply zcash_client_backend/src/fees/zip317.rs:353]: > Because of the output padding, this makes no difference to the fee of 15000 zat, and so the balance is calculated as 55000 - 40000 - 15000 = 0, i.e. we are in case 2, **but the flows are not transparent so this becomes case 3b**.

This is the bit of logic that was missing when I originally wrote this comment in the previous PR. I think this test is therefore fine as-is now.

Comment (daira) [reply zcash_client_backend/src/fees/zip317.rs]: I've implemented this now; please re-review.

File: zcash_client_backend/CHANGELOG.md [hunk 5]
```diff
@@ -20,6 +20,11 @@ and this library adheres to Rust's notion of
 
 ### Changed
 - MSRV is now 1.70.0.
+- `zcash_client_backend::{fixed,standard,zip317}::SingleOutputChangeStrategy`
+  now implement a slightly different strategy for choosing whether there will
+  be any change. Transactions that have any shielded non-change flows will
+  always have a change output (possibly zero-valued). This can avoid leaking
+  information about note amounts in some cases.
```
Comment (daira) [inline zcash_client_backend/CHANGELOG.md:27]: I will update this changelog entry.

File: zcash_client_sqlite/src/testing/pool.rs (hunk 4 above)
Comment (daira) [reply zcash_client_sqlite/src/testing/pool.rs:964]: This change is no longer necessary. I think it was because we were previously calculating the fee incorrectly.

File: zcash_client_sqlite/src/testing/pool.rs (hunk 3 above)
Comment (daira) [reply zcash_client_sqlite/src/testing/pool.rs:917]: This change is no longer necessary. I think it was because we were previously calculating the fee incorrectly.
